profiler/src/ProfilerEngine/Datadog.Profiler.Native.Linux/RingBuffer.cpp

```diff
@@ -0,0 +1,416 @@
+    _rb(rb)
+{
+}
+
+Buffer RingBuffer::Writer::Reserve(std::size_t size, bool* timeout) const
+{
+    std::size_t const n2 =
+        align_up(size + sizeof(RingBufferHeader), RingBufferAlignment);
+    if (n2 == 0)
+    {
+        return {};
+    }
+
+    auto* rb = _rb;
+    // \fixme{nsavoire} Not sure if spinlock is the best option here
+    std::unique_lock const lock{*rb->spinlock, ReserveTimeout};
+    if (!lock.owns_lock())
+    {
+        // timeout on lock
+        if (timeout)
+        {
+            *timeout = true;
+        }
+        return {};
+    }
+
+    // No need for atomic operation, since we hold the lock
+    std::uint64_t const writer_pos = *rb->writer_pos;
+
+    std::uint64_t const new_writer_pos = writer_pos + n2;
+
+    // Check that there is enough free space
+    std::uint64_t tail = __atomic_load_n(_rb->reader_pos, __ATOMIC_ACQUIRE);
+    if (rb->mask < new_writer_pos - tail)
+    {
+        return {};
+    }
+
+    uint64_t const head_linear = writer_pos & rb->mask;
+    auto* hdr =
+        reinterpret_cast<RingBufferHeader*>(rb->data + head_linear);
+
+    // Mark the sample as busy
+    hdr->size = size | RingBufferHeader::k_busy_bit;
+
+    // Atomic operation required to synchronize with reader load_acquire
+    __atomic_store_n(rb->writer_pos, new_writer_pos, __ATOMIC_RELEASE);
+
+    return {reinterpret_cast<std::byte*>(hdr + 1), size};
+}
+
+void RingBuffer::Writer::Commit(Buffer buf)
+{
+    auto* hdr = reinterpret_cast<RingBufferHeader*>(buf.data()) - 1;
+
+    // Clear busy bit
+    std::uint64_t new_size = hdr->size ^ RingBufferHeader::k_busy_bit;
+    // Needs release ordering to make sure that all previous writes are
+    // visible to the reader once reader acquires `hdr->size`.
+    __atomic_store_n(&hdr->size, new_size, __ATOMIC_RELEASE);
+}
+
+void RingBuffer::Writer::Discard(Buffer buf)
+{
+    auto* hdr = reinterpret_cast<RingBufferHeader*>(buf.data()) - 1;
+
+    // Clear busy bit
+    std::uint64_t new_size = hdr->size ^ RingBufferHeader::k_busy_bit;
+    new_size |= RingBufferHeader::k_discard_bit;
+    // Needs release ordering to make sure that all previous writes are
+    // visible to the reader once reader acquires `hdr->size`.
+    __atomic_store_n(&hdr->size, new_size, __ATOMIC_RELEASE);
+}
+
+RingBuffer::Reader::Reader(RingBufferImpl* rb) :
+    _rb(rb)
+{
+    _head = __atomic_load_n(_rb->writer_pos, __ATOMIC_ACQUIRE);
+    assert(rb->intermediate_reader_pos <= _head);
+    assert(_rb->intermediate_reader_pos == *_rb->reader_pos);
+}
+
+RingBuffer::Reader::~Reader()
+{
+    if (_rb && *_rb->reader_pos < _rb->intermediate_reader_pos)
+    {
+        __atomic_store_n(_rb->reader_pos, _rb->intermediate_reader_pos,
+                         __ATOMIC_RELEASE);
+    }
+}
+
+std::size_t RingBuffer::Reader::AvailableSamples(std::size_t sampleSize) const
+{
+    return (_head - _rb->intermediate_reader_pos) / (sizeof(RingBufferHeader) + sampleSize);
```

Comment:
I'm not sure this is correct if `sampleSize` is not aligned like what is done in `Writer.Reserve`:
`   n2 = align_up(size + sizeof(RingBufferHeader), RingBufferAlignment);`
This problem would disappear if the RingBuffer is created with the constant size of the buffers to store

Reply:
✔️